### DISCUSSION TITLE
Nim Lexer and Nim Multiline Comments

### DISCUSSION THREAD

**Discussion by sigzero** [orbitalquark/textadept#27] [General]
A multiline comment in Nim looks like this:

```
#[
      I am a comment!
]#
```
I tried the following in the `nim.lua` lexer but it doesn't work. Nim uses (IMO) a strange combo for the multiline comments.

local line_comment = lexer.to_eol('#', true)
local block_comment = lexer.range('#[', ']#')
lex:add_rule('comment', token(lexer.COMMENT, line_comment + block_comment))

Just wondering if it is even possible.

https://nim-by-example.github.io/comments/

**Comment by orbitalquark** [comment]
Try swapping the order:

    lex:add_rule('comment', token(lexer.COMMENT, block_comment + line_comment))

**Reply by sigzero** [reply]
Wow, that worked perfectly. Awesome.

**Comment by sigzero** [comment]
So looking at the NIm lexer the keywords etc. are out of date. The following URL seems to be the current version of things:

https://nim-lang.org/docs/system.html

Do I just copy everything under their respective heading from the page into the appropriate lexer section (avoiding things like \`ptr\` and super common keywords) etc.?

Is there a preferred lexer format? For example a section currently looks like this:

```
lex:add_rule('keyword', token(lexer.KEYWORD, word_match([[
  addr and as asm atomic bind block break case cast const continue converter
  discard distinct div do elif else end enum except export finally for from
  generic if import in include interface is isnot iterator lambda let macro
  method mixin mod nil not notin object of or out proc ptr raise ref return
  shared shl static template try tuple type var when while with without xor
  yield
]], true)))
```

Where I would write it like this to for better viewing and future updating:

```
lex:add_rule('keyword', token(lexer.KEYWORD, word_match([[
  addr and as asm atomic 
  bind block break 
  case cast const continue converter
  discard distinct div do 
  elif else end enum except export 
  finally for from
  generic 
  if import in include interface is isnot iterator 
  lambda let 
  macro method mixin mod 
  nil not notion 
  object of or out 
  proc ptr 
  raise ref return
  shared shl static 
  template try tuple type 
  var 
  when while with without 
  xor
  yield
]], true)))
```

**Comment by orbitalquark** [comment]
I would say to use your judgement on what to copy over for highlighting. I don't remember the source I used when writing the Nim lexer.

I think it's okay to format word match lists like you suggest. However, note that the Lua lexer uses comments to note which words were added in newer versions of the language. If Nim adds keywords in future versions, that may be worth noting similarly, and in that case the former block format is fine. You can use your judgement here, too.
